### PR TITLE
Addressed issue#2 Creation of ERC-721 adding requested extensions.

### PR DESCRIPTION
**Pull request by Sammmmx**
## Checklist

-  37 Tests Passing
-  No unrelated changes included
-  Educational .learn.md included

### CODE REVIEW

File: evm/nft-erc721/.gitignore [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+/cache
+/artifacts
+
+# TypeChain files
+/typechain
+/typechain-types
+
+# solidity-coverage files
+/coverage
+/coverage.json
```
Comment (PetarStoev02) [inline evm/nft-erc721/.gitignore:14]: Add `ignition/deployments/` to `.gitignore` — deployment artifacts are environment-specific and should not be committed. Same for `package-lock.json` if this is a template/recipe repo.

File: evm/nft-erc721/src/NFT721.sol [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import {ERC2981} from "@openzeppelin/contracts/token/common/ERC2981.sol";
+import {ERC721Enumerable} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+import {ERC721URIStorage} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+
+
+error NullAddress();
+error InvalidFee(uint256 _fee);
+error NoURI();
+
+//Deployed on Sepolia Testnet with contract address -> 0xC71C2E01AbC2FCc9F8E161aDe4C84A132be431d4
```
Comment (PetarStoev02) [inline evm/nft-erc721/src/NFT721.sol]: Remove this hardcoded deployment address from source code. Deployment info should live in the README or deployment scripts.

File: evm/nft-erc721/src/NFT721.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+contract NFT721 is ERC721, ERC721Enumerable, ERC721URIStorage, ERC2981, Ownable {
+    /// @dev Tracks the next token ID to be minted
+    uint256 private _nextTokenId;
+
+    /// @notice Emitted when default royalty is updated
+    /// @param _royaltyReceiver Address receiving royalty
+    /// @param _feeNumerator Royalty fee in basis points
```
Comment (PetarStoev02) [inline evm/nft-erc721/src/NFT721.sol:22]: Event names should follow PascalCase convention per Solidity style guide. Rename `_newDefaultRoyalty` → `DefaultRoyaltyUpdated`.

File: evm/nft-erc721/src/NFT721.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        public
+        onlyOwner
+        checkFee(feeNumerator)
+        returns (uint256)
+    {
+        if(bytes(uri).length == 0) revert NoURI();
+
+        uint256 tokenId = _nextTokenId++;
+        _safeMint(to, tokenId);
+        _setTokenURI(tokenId, uri);
```
Comment (PetarStoev02) [inline evm/nft-erc721/src/NFT721.sol:76]: When `royaltyReceiver` is `address(0)` but `feeNumerator > 0`, the fee is silently ignored. Consider reverting with `NullAddress()` if `feeNumerator > 0` and `royaltyReceiver == address(0)` to prevent accidental misconfiguration.

File: evm/nft-erc721/package.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "hardhat-project",
```
Comment (PetarStoev02) [inline evm/nft-erc721/package.json]: Rename from `"hardhat-project"` to `"nft-erc721"` to match the recipe name.

File: evm/nft-erc721/test/NFT721Test.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        .withArgs(alice.address);
+    });
+
+    it("should emit event _newDefaultRoyalty", async function () {
+      expect(await nft.setDefaultRoyalty(bob.address, "500"))
+        .to.emit(nft, "_newDefaultRoyalty")
+        .withArgs(bob.address, "500");
+    });
+  });
+
```
Comment (PetarStoev02) [inline evm/nft-erc721/test/NFT721Test.ts:153]: This event assertion is incorrect. `expect(await nft.setDefaultRoyalty(...))` resolves the tx before chai-matchers can intercept the event, so this test may pass vacuously. Use `await expect(nft.setDefaultRoyalty(...)).to.emit(...)` instead (no `await` inside `expect`).

**Review by PetarStoev02 — CHANGES_REQUESTED** [review]
Good work on the contract and tests. A few things need to be addressed before merging.

**Blocker:** The entire `ignition/deployments/chain-11155111/` directory (~136k lines including build-info) and `package-lock.json` (~8k lines) should be removed from this PR and added to `.gitignore`. These are environment-specific artifacts that shouldn't be in the repo.

File: evm/nft-erc721/src/NFT721.sol (hunk 4 above)
Comment (Sammmmx) [reply evm/nft-erc721/src/NFT721.sol:76]: You are right, fee is silently ignored. I have applied a revert if fee is added for empty royaltyReceiver, moreover I believe the fee should not be 0 if a royaltyReceiver is added, am I right?

Comment (Sammmmx) [reply evm/nft-erc721/src/NFT721.sol:76]: Because if a royaltyReceiver is added then the motive should be for that NFT Creator to receive the fee.

File: evm/nft-erc721/.gitignore (hunk 1 above)
Comment (Sammmmx) [reply evm/nft-erc721/.gitignore:14]: Added ignition/deployments and package-lock.json to .gitignore

File: evm/nft-erc721/package.json (hunk 5 above)
Comment (Sammmmx) [reply evm/nft-erc721/package.json]: Renamed to nft-erc721

File: evm/nft-erc721/test/NFT721Test.ts (hunk 6 above)
Comment (Sammmmx) [reply evm/nft-erc721/test/NFT721Test.ts:153]: changed it to await expect() while updating event name.

File: evm/nft-erc721/src/NFT721.sol (hunk 3 above)
Comment (Sammmmx) [reply evm/nft-erc721/src/NFT721.sol:22]: _newDefaultRoyalty is now changed to DefaultRoyaltyUpdated.

File: evm/nft-erc721/src/NFT721.sol (hunk 2 above)
Comment (Sammmmx) [reply evm/nft-erc721/src/NFT721.sol]: Removed the Testnet contract address.

**Comment by Sammmmx** [comment]
Apologies, due to a mix up in forks and branches I had pushed changes in the wrong place. 
Requesting a review.

**Review by PetarStoev02 — CHANGES_REQUESTED** [review]
Thanks for pushing the fixes @Sammmmx — most of the original items are resolved. Nice work on the royalty validation logic and the event fix.

**Status of previous review items:**
- ✅ Hardcoded deployment address removed from source
- ✅ Event renamed to PascalCase (`DefaultRoyaltyUpdate`)
- ✅ `await expect()` pattern fixed for event assertion
- ✅ `NullAddress()` revert added when `royaltyReceiver == address(0)` with `feeNumerator > 0`
- ✅ `package.json` name updated to `nft-erc721`
- ✅ `.gitignore` updated with `ignition/deployments`

**Remaining issues:**

### Blocking

1. **`ignition/deployments/` is still committed** — the `.gitignore` change only prevents *future* tracking, but the directory is already in this PR (136k+ lines of build artifacts). You need to remove the tracked files:
   ```bash
   git rm -r --cached evm/nft-erc721/ignition/deployments
   git commit -m "Remove committed deployment artifacts"
   ```

2. **README.md is the default Hardhat boilerplate** — it references "Sample Hardhat Project" and `Lock.ts`. Replace with a proper README for the nft-erc721 recipe. See `evm/token-erc20/` for reference.

### Non-blocking (fix in this PR or follow-up)

3. **`package-lock.json` should NOT be in `.gitignore`** — the [TEMPLATE_STANDARD.md](../TEMPLATE_STANDARD.md) requires lockfiles to be committed for reproducible installs. Remove `package-lock.json` from `.gitignore`. *(My original review was wrong about this — apologies.)*

4. **`.env.example` should use empty values** per the template standard:
   ```
   PRIVATE_KEY=
   ALCHEMY_URL=
   ETHERSCAN_API_KEY=
   ```

5. **Missing newline at EOF** in `.gitignore` and `.env.example` — template standard requires all files to end with a newline.

6. **`.learn.md` is missing a dedicated "Security Considerations" section** — the template standard requires security considerations specific to this contract (e.g., owner-only minting, royalty bypass on non-compliant marketplaces, no burn mechanism).

---

**Re: your question about `feeNumerator == 0` with a non-zero `royaltyReceiver`** — allowing it is fine. A creator might want to register a receiver now with 0% fee and update it later via `setDefaultRoyalty`. No need to revert on that case.

**Comment by PetarStoev02** [comment]
Hey @Sammmmx, I've invited you to the w3-kit org — you should have a pending invite in your GitHub notifications. Once you accept, you can clone the repo directly and push branches without needing a fork. Should make the PR workflow much easier going forward.

**Review by PetarStoev02 — APPROVED** [review]
All blockers from the previous review are resolved — nice work @Sammmmx.

A few minor items remain that we'll handle in a follow-up commit:
- README typo (`igniton` → `ignition`) + trailing empty code fences
- `package-lock.json` root name mismatch (`hardhat-project` → `nft-erc721`)
- Missing test for null-address royalty guard path
- Dead code in `hardhat.config.ts`
- `.gitignore` typo (`Deployement` → `Deployment`)
- Spacing after `if` keyword in contract
